main.cpp

```diff
@@ -1,95 +1,170 @@
+#include <stdexcept>
+#include <utility>
+#include <type_traits>
+#include <iterator>
+#include <cstddef>
 
+template <typename T>
 struct Node {
     // 这两个指针会造成什么问题？请修复
-    std::shared_ptr<Node> next;
-    std::shared_ptr<Node> prev;
+    // 对于双向链表，相邻项之间存在循环引用，引用计数无法归零释放空间。
+    // 对于双向链表，可以看作前一项“拥有”后一项。
+    // 而后一项保留前一项的Node*指针
+    std::unique_ptr<Node> next;
+    Node* prev;
     // 如果能改成 unique_ptr 就更好了!
 
-    int value;
-
-    // 这个构造函数有什么可以改进的？
-    Node(int val) {
-        value = val;
-    }
-
-    void insert(int val) {
-        auto node = std::make_shared<Node>(val);
-        node->next = next;
-        node->prev = prev;
-        if (prev)
-            prev->next = node;
-        if (next)
-            next->prev = node;
-    }
+    T value;
+
+    // 这个构造函数有什么可以改进的？:value直接根据val构造而不是默认构造后赋值。
+    Node(const T& val): value(val), prev(nullptr) {}
+    Node(const Node&) = default;
+    Node& operator=(Node&&) = default;
+
+    // insert会导致无法使用unique_ptr，因为会破环上面假设的“前一项拥有后一项”的前提
+    /*
+              +--- O ---+
+        O ---x           x--- nextO
+              +--- O ---+
+
+        会变成上面这样, 双向链表中用不到该操作所以直接注掉了。
+    */
+    // void insert(int val) {
+    //     auto node = std::make_unique<Node>(val);
+    //     node->next = next;
+    //     node->prev = prev;
+    //     if (prev)
+    //         prev->next = node;
+    //     if (next)
+    //         next->prev = node;
+    // }
 
     void erase() {
-        if (prev)
-            prev->next = next;
         if (next)
             next->prev = prev;
+        if (prev)
+            prev->next = std::move(next);
     }
 
     ~Node() {
-        printf("~Node()\n");   // 应输出多少次？为什么少了？
+        printf("~Node()\n");   // 应输出多少次？为什么少了？因为循环引用
     }
 };
 
+template<typename T>
 struct List {
-    std::shared_ptr<Node> head;
+
+    class iterator {
+    public:
+        using iterator_category = std::bidirectional_iterator_tag;
+        using difference_type   = std::ptrdiff_t;
+        using value_type        = Node<T>;
+        using pointer           = value_type*;  // or also value_type*
+        using reference         = value_type&;  // or also value_type&
+
+        iterator(pointer ptr): m_ptr(ptr) {}
+        reference operator*() { return *m_ptr; }
+        pointer operator->() { return m_ptr; }
+
+        // Prefix increment
+        iterator& operator++() 
+        { m_ptr = m_ptr->next.get(); return *this; }  
+
+        // Postfix increment
+        iterator operator++(int) 
+        { iterator tmp = *this; ++(*this); return tmp; }
+
+        friend bool operator== (const iterator& a, const iterator& b) 
+        { return a.m_ptr == b.m_ptr; };
+        friend bool operator!= (const iterator& a, const iterator& b) 
+        { return a.m_ptr != b.m_ptr; };     
+
+    private:
+        Node<T>* m_ptr;
+    };
+
+    std::unique_ptr<Node<T>> head;
+    Node<T>* back = nullptr;
 
     List() = default;
 
-    List(List const &other) {
+    List(List &other) {
```

Comment:
```suggestion
    List(List const &other_) {
       List &other = const_cast<List &>(other_);
```
拷贝构造函数有特定的函数签名，不建议轻易改变。